xds/src/main/java/io/grpc/xds/XdsNameResolver2.java

```diff
@@ -0,0 +1,378 @@
+      final String finalCluster = cluster;
+      class SelectionCompleted implements Runnable {
+        @Override
+        public void run() {
+          releaseCluster(finalCluster);
+        }
+      }
+
+      Map<String, ?> rawServiceConfig =
+          generateServiceConfigWithMethodConfig(
+              args.getMethodDescriptor().getFullMethodName(),
+              selectedRoute.getRouteAction().getTimeoutNano());
+      if (logger.isLoggable(XdsLogLevel.INFO)) {
+        logger.log(XdsLogLevel.INFO,
+            "Generated service config (method config):\n{0}", new Gson().toJson(rawServiceConfig));
+      }
+      ConfigOrError parsedServiceConfig = serviceConfigParser.parseServiceConfig(rawServiceConfig);
+      Object config = parsedServiceConfig.getConfig();
+      if (config == null) {
+        throw new AssertionError(
+            "Bug: invalid config", parsedServiceConfig.getError().asException());
+      }
+      return
+          Result.newBuilder()
+              .setCallOptions(args.getCallOptions().withOption(CLUSTER_SELECTION_KEY, cluster))
+              .setConfig(config)
+              .setCommittedCallback(new SelectionCompleted())
+              .build();
+    }
+
+    private boolean retainCluster(String cluster) {
+      AtomicInteger refCount = clusterRefs.get(cluster);
+      if (refCount == null) {
+        return false;
+      }
+      int count;
+      do {
+        count = refCount.get();
+        if (count == 0) {
+          return false;
+        }
+      } while (!refCount.compareAndSet(count, count + 1));
+      return true;
+    }
+
+    private void releaseCluster(String cluster) {
+      int count = clusterRefs.get(cluster).decrementAndGet();
+      if (count == 0) {
+        clusterRefs.remove(cluster);
+        syncContext.execute(new Runnable() {
+          @Override
+          public void run() {
+            updateResolutionResult();
+          }
+        });
+      }
+    }
+  }
+
+  private void updateResolutionResult() {
+    Map<String, ?> rawServiceConfig =
+        generateServiceConfigWithLoadBalancingConfig(clusterRefs.keySet());
+    if (logger.isLoggable(XdsLogLevel.INFO)) {
+      logger.log(
+          XdsLogLevel.INFO, "Generated service config:\n{0}", new Gson().toJson(rawServiceConfig));
+    }
+    ConfigOrError parsedServiceConfig = serviceConfigParser.parseServiceConfig(rawServiceConfig);
+    Attributes attrs =
+        Attributes.newBuilder()
+            .set(XdsAttributes.XDS_CLIENT_POOL, xdsClientPool)
+            .set(InternalConfigSelector.KEY, new ConfigSelector())
+            .build();
+    ResolutionResult result =
+        ResolutionResult.newBuilder()
+            .setAttributes(attrs)
+            .setServiceConfig(parsedServiceConfig)
+            .build();
+    listener.onResult(result);
+  }
+
+  @SuppressWarnings("ModifyCollectionInEnhancedForLoop")  // ok for concurrent map
+  private class ConfigWatcherImpl implements ConfigWatcher {
```

Comment:
Will address and format classes/methods correctly after all the other comments are addresses (just to avoid moving blocks of code around and make the review hard).